### DISCUSSION TITLE
[GtkSharp] Error: Several functions Run can be called

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3006] [Q&A] [answered]
Install nuget package `GtkSharp` and try with this piece of code:

```
uses Gtk;

begin
  Gtk.Application.Init();
  var window := new Gtk.Window('helloworld');
  window.Show();
  Gtk.Application.Run();
end.
```

I tried my best without success. You can see I explicitly use `Gtk.` to avoid conflicting with `WinForms`. I even started with a `Console Application` project to get rid of any references to `WinForms`. It still failed with exactly this very same error.

**Comment by SunSerega** [comment]
The error should attach itself to a particular function - which is it, and what other overloads does it have?

**Reply by ghost** [reply]
I don't know! This is the reason I asked for help here. I loaded `GtkSharp.dll` into `ILSpy`, I can only found just one version of `Gtk.Application.Run()`. I have no idea where other overloads are.

**Reply by SunSerega** [reply]
Ok, so you are saying you do know it is `Gtk.Application.Run`? (P.S. ok, now I see `Run` in the discussion title)

Overloads are:
```
##
procedure p1(x: integer) := TypeName(x).Println;
procedure p1(x: string) := TypeName(x).Println;

p1(1);
p1('abc');
```
Basically, different subprogram signatures with the same name. The right subprogram is chosen based on input types.
If there are multiple options:
```
##
procedure p1(a: integer; b: real) := exit;
procedure p1(a: real; b: integer) := exit;

p1(1,2);
```
It can either call `p1(1, 2.0)` or `p1(1.0, 2)`. Since there is no better/worse choice, compiler just says that multiple subprograms can be called.

When you place opening bracket (the `(` symbol) after a subprogram name, it shows the description, signature, and up/down buttons (can be activated using keyboard) to cycle through all the overloads.

**Reply by SunSerega** [reply]
That said, I looked at GtkSharp source (`gtk\Application.cs`) are there is in fact only one `Run` function... Maybe it's actually some kind of compiler bug.
Please make a zip (or rar or whatever) with the minimum amount of code and .dll-s required to reproduce this, I'll try to debug the compiler.

**Reply by ghost** [reply]
[Project7.zip](https://github.com/pascalabcnet/pascalabcnet/files/13918632/Project7.zip)

Here you are.

**Reply by SunSerega** [reply]
Yeah, something's really broken in there... The compiler is also trying to call `GLib.Application.Run()`. Which is not only in a different namespace but also is an instance function (you cannot call it without first calling `new`)

But it might not even be a compiler bug. Look:
```
  var t := typeof(Gtk.Application);
  Println(t);
  t.GetMethods
  .Where(mi->mi.Name='Run')
  .PrintLines(mi->$'{mi.DeclaringType} : {mi}');
```
For me this outputs:
```
Gtk.Application 
Gtk.Application : Void Run()
GLib.Application : Int32 Run()
GLib.Application : Int32 Run(System.String, System.String[])
```
So CLR somehow finds methods from a different type from a different namespace???

Can probably be hot-fixed in the Pascal compiler, but first need to investigate deeper...

**Reply by SunSerega** [reply, marked as answer]
Ah, wait, no, `Gtk.Application` is derived from `GLib.Application`.
So the problems:
1. `Gtk.Application.Run` fails to hide inherited method `GLib.Application.Run`
2. `GLib.Application.Run` is still considered valid, despite being an instance method.

(both 100% pascal compiler bugs)

**Reply by SunSerega** [reply]
https://github.com/pascalabcnet/pascalabcnet/issues/3008

As a work-around for now, you can use reflection during runtime:
```
uses Gtk;

begin
  Gtk.Application.Init();
  var window := new Gtk.Window('helloworld');
  window.Show();
  
  typeof(Gtk.Application).GetMethods
  .Where(mi->mi.Name='Run')
  .Single(mi->mi.DeclaringType=typeof(Gtk.Application))
  .Invoke(nil, nil);
  
end.
```
